content/docs/specs/remote_write_spec_2_0.md

```diff
@@ -130,18 +130,42 @@ Senders MUST include a user agent header that SHOULD follow [the RFC 9110 User-A
 
 ### Response
 
-Receivers ingesting all samples successfully MUST return a 200 HTTP status code. In such a successful case, the response body from the Receiver SHOULD be empty; Senders MUST ignore the response body. The response body is RESERVED for future use.
+Receivers ingesting all Samples and Histograms successfully MUST return a [success 2xx HTTP status code](https://www.rfc-editor.org/rfc/rfc9110.html#name-successful-2xx). In such a successful case, the response body from the Receiver SHOULD be empty and the status code SHOULD be [204 HTTP No Content](https://www.rfc-editor.org/rfc/rfc9110.html#name-204-no-content); Senders MUST ignore the response body. The response body is RESERVED for future use.
 
-Receivers MUST NOT return a 200 HTTP status code if any of the samples were not written successfully (e.g. on a [partial write](#partial-write) or a full write rejection). In such a case, the Receiver MUST provide a human-readable error message in the response body. The Receiver's error SHOULD contain information about the amount of the samples being rejected and for what reasons. Senders MUST NOT try and interpret the error message and SHOULD log it as is.
+Receivers MUST NOT return a 2xx HTTP status code if any of the Samples or Histograms were NOT written successfully (e.g. on a [partial write](#partial-write) or a full write rejection). In such a case, the Receiver MUST provide a human-readable error message in the response body. The Receiver's error SHOULD contain information about the amount of the samples being rejected and for what reasons. Senders MUST NOT try and interpret the error message and SHOULD log it as is.
 
-The following subsections specify Sender and Receiver semantics around different write error cases.
+The following subsections specify Sender and Receiver semantics around headers and different write error cases.
+
+#### Response Headers
+
+<!---
+Rationales: https://github.com/prometheus/prometheus/issues/14359
+-->
+Upon a successful content negotiation, Receivers ingest the received Samples, Histograms and Exemplars. In such cases, Receivers MUST track the total number of Samples, Histograms or Exemplars that were accepted or ingested. Receivers MUST share those statistics with the Sender as the HTTP response `X-Prometheus-Remote-Write-Accepted-*` headers specified below. Receivers MAY omit those when the value is zero.
```

Comment:
> Receivers MAY omit those when the value is zero.

I don't think we should allow this. It's yet another way to confuse the end user. Did I get no values because nothing was accepted, or because I sent 2.0 and the Receiver only speaks 1.0?

Reply:
Great question and I'm glad you challenged this! Was removing and adding this back like 5 times. It's not clear what will confuse "user" (I assume you mean Sender) here (: 

TL;DR I see the Pros of this, but not sure yet, let's chat on our sync.

Let's unpack:

**What if we say it's MUST for Receiver to put those headers even if it's 0.**

**Pros:**
* For 2xx it's indeed clear if we talk to Receiver 1.0 or 2.0 (potentially 2.1 in future too). This helps a bit on 2 cases below:
  a. It simplifies the known problem detection for 2xx a lot e.g. used to be 

    > you got 2xx AND (accepted samples == 0 and/or accepted histograms == 0) AND you sent some samples/histograms == you talk to 1.0 receiver with known bug of not checking content type.

    Now, it's more like "you got 2xx and none of the required headers in 2.0 == that case". 

  b. Allows us actually to check if 2.0 Receiver correctly uses 1.0 payload we sent. Had to ignore this case in https://github.com/prometheus/prometheus/pull/14444 as you don't know what Receiver version it is.

I don't see any other benefits yet. Particularly on any non 2xx case Sender cannot assume the Receiver spec version (see cons why).

**Cons:**
* Sender implementation complexity. Treating empty and zero value as different cases is generally asking for complexity, so wanted to avoid this. In Go it sucks to represent this, you are force to use pointers or flags (`hasSpecified`). It's the reason why it sucks sometimes to work with proto or yamls or any data format and we end up with those `aws.String` or `proto.Int` that just pointerize for little reason. In other languages it requires more bytes to store and complex data structures too. If we can avoid it, it would be amazing.
* Sender confusion due to inconsistency with other failure use cases. Zero default is consistent with thousands of cases where you will NOT have those headers because e.g. HTTP didn't even hit Receiver (e.g. proxy errors, initial HTTP validation before parsing proto). This means we can't rely on this behaviour for non 2xx.
* Wanted to avoid relying on Sender doing special things for certain spec version as long as possible. Let's list why it's bad/not ideal, especially for failure cases:
  * Similar to above, we will have cases of non-2xx AND No headers EVEN for 2.0 Receivers. So now Senders can't fully assume Receiver spec version on ANY error case anyway, but they might be tempted too.
  * Receiver Implementation now need to ask themselves "do I need to add those headers the moment I start processing the request or only after I decode the proto or should my forward proxy do that? How far should I unsure this". It will pain to even define this, I tried with "when processing decoded proto" mention.
* Inconsistency with forward compatibility of 2.0 Receiver. If we add new header in 2.1 (perhaps a new type of sample) the 2.0 receivers would NOT add them. However such new sample does not exists, so it's definitely means zero, which offers a clean Sender implementation. With required 0 header, Senders now require more friction to implement as `X-Prometheus-Remote-Write-Accepted-SomethingAddedIn2_1` now holds two meaning, but this information is useless it all means just zero after all.
  * Counter argument: This only works if we can guarantee future headers omission will mean zero. One tricky example is e.g. metadata header in future, so can we guarantee that?

Reply:
Another question - what Sender should do with formatted response header (e.g. `X-Prometheus-Remote-Write-Accepted-Samples 12.44-yolo`)?

Reply:
We found some consensus in https://cloud-native.slack.com/archives/C06B8KJUU2Y/p1720628447889389 will try to put that on the PR.